plugins/modules/s3_bucket_notification.py

```diff
@@ -221,36 +338,57 @@ def main():
         lambda_version=dict(type='int', default=0),
     )
 
+    mutually_exclusive = [
+        ['queue_arn', 'topic_arn', 'lambda_function_arn'],
+        ['lambda_alias', 'lambda_version']
+    ]
+
     module = AnsibleAWSModule(
         argument_spec=argument_spec,
         supports_check_mode=True,
-        mutually_exclusive=[['lambda_alias', 'lambda_version']],
+        mutually_exclusive=mutually_exclusive,
         required_if=[['state', 'present', ['events']]]
     )
 
-    bucket = AmazonBucket(module.client('s3'), module.params['bucket_name'])
+    client = module.client('s3')
+    bucket = AmazonBucket(module, client)
     current = bucket.current_config(module.params['event_name'])
     desired = Config.from_params(**module.params)
-    notification_configuration = [cfg.raw for cfg in bucket.full_config()]
+
+    notification_configs = dict(
+        QueueConfigurations=[],
+        TopicConfigurations=[],
+        LambdaFunctionConfigurations=[]
+    )
+
+    for target_configs in bucket.full_config():
+        for cfg in bucket.full_config()[target_configs]:
+            notification_configs[target_configs].append(camel_dict_to_snake_dict(cfg.raw))
 
     state = module.params['state']
-    try:
-        if (state == 'present' and current == desired) or (state == 'absent' and not current):
-            changed = False
-        elif module.check_mode:
+    updated_configuration = dict()
+    changed = False
+
+    if state == 'present':
+        if current != desired:
+            updated_configuration = bucket.apply_config(desired)
             changed = True
-        elif state == 'present':
+        if module.check_mode:
```

Comment:
Do we really need this check here since check_mode seems to be already handled inside apply_config() through _upload_bucket_config()?